gradio/utils.py

```diff
@@ -639,8 +639,8 @@ async def asyncgen_wrapper(*args, **kwargs):
                 before_fn(*before_args)
             async for response in f(*args, **kwargs):
                 yield response
-            if after_fn:
-                after_fn(*after_args)
+                if after_fn:
```

Comment:
I don't understand - why should we call the after_fn (which in this case sets in_event_listener to false) right after the first yield, if there are more responses left to be yielded? Why wouldn't we call it after we have yielded all the responses?

Reply:
We need to do this because we create components during the postprocessing (after each iteration in the case of generators) and we need `in_event_listener` to be False for the component to be properly created. 

This is the relevant code -

```python
  state[output_id] = self.blocks[output_id].__class__(**kwargs)

  prediction_value = postprocess_update_dict(
      block=state[output_id],
      update_dict=prediction_value,
      postprocess=block_fn.postprocess,
  )
```

Right now `in_event_listener` is always True because the after function never runs.

Reply:
Ok this makes sense to me. I believe in this case we should be calling the before_fn before every function call, and calling after_fn right after every function call, but before the yield. Something like (haven't tested this):
```
        @functools.wraps(f)
        async def asyncgen_wrapper(*args, **kwargs):
            while True:
                if before_fn:
                    before_fn(*before_args)
                try:
                    response = next(f(*args, **kwargs))
                except StopAsyncIteration:
                    if after_fn:
                        after_fn(*after_args)
                    break
                if after_fn:
                    after_fn(*after_args)
                yield response
```

This way, in_event_listener is set properly to only be applied while the function is running

Reply:
Nice suggestion! I took a separate approach but the end result is the same, we call before_fn and after_fn for each iteration. I also realized this closes #6762!